test/router.js

```diff
@@ -138,6 +139,131 @@ test('Should close on unregistered path', t => {
   })
 })
 
+test('Should call global handler on unregistered path', t => {
+  t.plan(3)
+  const fastify = Fastify()
+
+  t.tearDown(() => fastify.close())
+
+  fastify.register(fastifyWebsocket, {
+    handle: (conn, request) => {
+      conn.socket.on('message', message => {
+        try {
+          conn.socket.send('hi from global handler')
+        } catch (err) {
+          conn.socket.send(err.message)
+        }
+      })
+
+      t.tearDown(conn.destroy.bind(conn))
+    }
+  })
+
+  fastify.get('/echo', { websocket: true }, (conn, request) => {
+    conn.socket.on('message', message => {
+      try {
+        conn.socket.send('hi from /echo handler')
+      } catch (err) {
+        conn.socket.send(err.message)
+      }
+    })
+
+    t.tearDown(conn.destroy.bind(conn))
+  })
+
+  fastify.listen(0, err => {
+    t.error(err)
+    const ws = new WebSocket('ws://localhost:' + (fastify.server.address()).port)
+    const client = WebSocket.createWebSocketStream(ws, { encoding: 'utf8' })
+    t.tearDown(client.destroy.bind(client))
+
+    ws.on('open', () => {
+      ws.send('hi from client')
+      client.end()
+    })
+
+    ws.on('message', message => {
+      t.equal(message, 'hi from global handler')
+    })
+
+    ws.on('close', () => {
+      t.pass()
+    })
+  })
+})
+
+test('Should invoke the correct handler depending on the headers', t => {
+  const fastify = Fastify()
+
+  t.tearDown(() => fastify.close())
+
+  fastify.register(fastifyWebsocket)
+  fastify.route({
+    method: 'GET',
+    url: '/',
+    handler: (request, reply) => {
+      reply.send('hi from handler')
+    },
+    wsHandler: (conn, request) => {
+      conn.write('hi from wsHandler')
+      t.tearDown(conn.destroy.bind(conn))
+    }
+  })
+
+  fastify.listen(0, err => {
+    t.error(err)
+
+    const port = fastify.server.address().port
+
+    const httpClient = net.createConnection({ port: port }, () => {
+      httpClient.write('GET / HTTP/1.1\r\n\r\n')
+      httpClient.once('data', data => {
+        t.match(data.toString(), /hi from handler/i)
+      })
+    })
+
+    const wsClient = net.createConnection({ port: port }, () => {
+      wsClient.write('GET / HTTP/1.1\r\nConnection: upgrade\r\nUpgrade: websocket\r\nSec-WebSocket-Key: dGhlIHNhbXBsZSBub25jZQ==\r\nSec-WebSocket-Version: 13\r\n\r\n')
+      wsClient.once('data', data => {
+        t.match(data.toString(), /hi from wsHandler/i)
+        wsClient.end(() => { t.end() })
+      })
+    })
+  })
+})
+
+test('Should call the global handler if a non-websocket route with path exists', t => {
+  const fastify = Fastify()
+  t.tearDown(() => fastify.close())
+
+  fastify.get('/http', (request, reply) => {
+    t.ok('Should not call http handler')
```

Comment:
@mcollina, this test normally should fail (I changed it to pass just so I can push)
We already have an http route registered with the same path, so the router then matches it and returns the http handler which has no information about the websocket connection.

This is the issue I was referring to the other day caused by having the global handler operate the way it does right now.

Reply:
An interesting solution to keep the global handler here would be to use the `onRoute` hook to wrap all handlers (not just the ones with websocket handlers defined) and make them call the global handler in case it's a websocket connection and they have no websocket handler defined.

```js
routeOptions.handler = (request, reply) => {
    if (request.raw[kWs]) {
        reply.hijack()
        // wsHandler can be either the route handler or global handler
        const result = wsHandler.call(fastify, request.raw[kWs], request)
        if (result && typeof result.catch === 'function') {
            result.catch(err => request.raw[kWs].destroy(err))
        }
    } else {
        return handler.call(fastify, request, reply)
    }
}
```

With this and the `defaultRoute` updated, we can cover all use cases, the downside would be the overhead an additional function call and condition check for each handler call, but I think that's worth it if we want to keep the global handler. 

What do you think @mcollina?

Reply:
I implemented this solution, it works if the user registers all routes after the `fastify-websocket` plugin is registered, otherwise the `onRoute` hook won't be called for the previous routes and the handler can't be updated to call the global handler in case of a match.

Also, the global handler is called in 2 scenarios:
1/ A route with no websocket handler is matched, the global handler is called instead.
2/ No route match is found, the updated `defaultRoute` is called

This means that for `1` the hooks are executed and we go through the normal fastify request cycle before running the global handler, but for `2` we don't.

I've tried hard to find a solution that would allow us to keep the global handler and cover all use cases, but this is the best one I was able to come up with, if it's acceptable to document this behavior in the Readme file we'll be able to keep the global handler, otherwise we should discuss alternatives.

Let me know what you think

Reply:
@mcollina what do you think about this? 
I've updated the docs as well